<!doctype html>
<head>
    <title>IEU OpenGWAS API</title>

    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">

    <link rel="apple-touch-icon" sizes="180x180" href="{{ url_for('static', filename='images/logo/opengwas/favicon/apple-touch-icon.png') }}">
    <link rel="icon" type="image/png" sizes="32x32" href="{{ url_for('static', filename='images/logo/opengwas/favicon/favicon-32x32.png') }}">
    <link rel="icon" type="image/png" sizes="16x16" href="{{ url_for('static', filename='images/logo/opengwas/favicon/favicon-16x16.png') }}">
    <link rel="manifest" href="{{ url_for('static', filename='images/logo/opengwas/favicon/site.webmanifest') }}" crossorigin="use-credentials">
    <link rel="mask-icon" href="{{ url_for('static', filename='images/logo/opengwas/favicon/safari-pinned-tab.svg') }}" color="#5bbad5">
    <link rel="shortcut icon" href="{{ url_for('static', filename='images/logo/opengwas/favicon/favicon.ico') }}">
    <meta name="msapplication-TileColor" content="#da532c">
    <meta name="msapplication-config" content="{{ url_for('static', filename='images/logo/opengwas/favicon/browserconfig.xml') }}">
    <meta name="theme-color" content="#ffffff">

    <!--bootstrap darkly theme-->
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/4.6.2/darkly/bootstrap.min.css" integrity="sha512-nIedR8ngz58VrJKgdNSou8ny5bcqiIbZXA8YQPjK4whu6U6o8wGbP4pJSLBZfrEa77FA//cDy+1shdhy53QvBg==" crossorigin="anonymous" referrerpolicy="no-referrer" />
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/tocbot/4.23.0/tocbot.css" integrity="sha512-MiKzJ1SHUjeQpWOIKdj1LYmEdnPF076Zeq7en/98oFIfomw5MEbqJ2RhRl+4WHo1ciJyFjFigtgKzXOO50Psrg==" crossorigin="anonymous" referrerpolicy="no-referrer" />

</head>

<body data-spy="scroll" data-target="#toc">

    <main>

        <div class="container">

            <div class="row">
                <div class="col-md-3"></div>
                <div class="col-md-9 text-right">
                    <img src="{{ url_for('static', filename='images/logo/opengwas/text-white.png')}}" height="100" style="margin-bottom: 30px;">
                </div>
            </div>

            <div class="row">
                <div class="col-md-3">
                    <div class="toc js-toc sticky-top"></div>
                </div>
                <div class="col-md-9 js-toc-content">
                    <section id="heading">
                        <h3>API status</h3>
                        <div class="jumbotron" style="padding: 20px;">
                            <div class="row">
                                <div class="col-md-6">
                                    <ul class="list-group" style="list-style: none;">
                                        <li class=" "><span class="badge badge-success">{{ status['API version'] }}</span> Version</li>
                                        <li class=" "><span class="badge badge-{{ 'success' if status['Access'] == 'public' else 'danger' }}">{{ status['Access'] }}</span> Access level</li>
                                    </ul>
                                    <p>
                                        {% if elastic_counts %}
                                            <span class="badge badge-success">{{ "{:,}".format(elastic_counts) }}</span> genetic associations<br/>
                                        {% endif %}
                                        {% if neo4j_counts %}
                                            <span class="badge badge-success">{{ "{:,}".format(neo4j_counts) }}</span> GWAS summary datasets.
                                        {% endif %}
                                    </p>
                                </div>
                                <div class="col-md-6">
                                    <ul class="list-group" style="list-style: none;">
                                        <li class=" "><span class="badge badge-{{ 'success' if status['Neo4j'] == 'Available' else 'danger' }}">{{ status['Neo4j'] }}</span> Neo4j status</li>
                                        <li class=" "><span class="badge badge-{{ 'success' if status['ElasticSearch'] == 'Available' else 'danger' }}">{{ status['ElasticSearch'] }}</span> ElasticSearch status</li>
                                        <li class=" "><span class="badge badge-{{ 'success' if status['LD reference panel'] == 'Available' else 'danger' }}">{{ status['LD reference panel'] }}</span> LD reference panel</li>
                                        <li class=" "><span class="badge badge-{{ 'success' if status['PLINK executable'] == 'Available' else 'danger' }}">{{ status['PLINK executable'] }}</span> PLINK executable</li>
                                        <li class=" "><span class="badge badge-{{ 'success' if status['Airflow'] == 'Available' else 'danger' }}">{{ status['Airflow'] }}</span> Airflow</li>
                                        <li class=" "><span class="badge badge-{{ 'success' if status['1000 genomes annotation VCF'] == 'Available' else 'danger' }}">{{ status['1000 genomes annotation VCF'] }}</span> Population LD scores and frequencies</li>
                                    </ul>
                                </div>
                            </div>
                        </div>
                        <div class="row">
                            <div class="col">
                                <p class="lead text-left">
                                    <a class="btn btn-warning btn-lg" href="{{ url_for('api.doc') }}" target="_blank" role="button"><span class="glyphicon glyphicon-star"></span>Documentation</a>
                                    {% if status['Access'] == 'private' %}
                                        <a class="btn btn-warning btn-lg" href="https://ieugit-scmv-d0.epi.bris.ac.uk/ml18692/igd-upload-docs" role="button">Upload instructions</a>
                                    {% endif %}
                                    <a class="btn btn-secondary btn-lg" href="https://gwas.mrcieu.ac.uk" role="button">GWAS home page</a>
                                </p>
                            </div>
                        </div>
                    </section>

                    <hr class='my-4'>

                    <section>
                        <h3 id="overview">Overview</h3>
                        <p>This is a <abbr title="Representational state transfer application programming interface">REST</abbr>ful <abbr title="Application programming interface">API</abbr> for programmatic access to the <abbr title="MRC Integrative Epidemiology Unit">IEU</abbr> Open<abbr title="Genome wide association study">GWAS</abbr> database. The API is designed to support the following operations:</p>
                        <ul>
                            <li>Make available study information about all the GWAS datasets in the database</li>
                            <li>Obtain the top hits from a GWAS, with optional automatic clumping.</li>
                            <li>Find any set of variant-trait associations based on rsid or chromosome:position.</li>
                            <li>Automatically find <abbr title="linkage disequilibrium">LD</abbr> proxies if a variant is missing from a GWAS database.</li>
                            <li>Calculate LD matrices across a set of variants.</li>
                            <li>Phenome-wide association study (PheWAS) - Perform fast lookups of a particular variant across all datasets.</li>
                            <li>Get <a href="https://www.ncbi.nlm.nih.gov/snp/" target="_blank">dbSNP</a> information and annotations on a set of variants.</li>
                        </ul>
                        <p>The API can be used directly by your applications, or accessed through convenient packages in R and Python. More information on those below. The <a href="https://github.com/MRCIEU/TwoSampleMR" target="_blank">TwoSampleMR</a> R package, and <a href="http://www.mrbase.org/" target="_blank">MR-Base</a> web apps both use this API in order to automatically perform Mendelian randomization for causal inference. The <a href="https://mrcieu.github.io/gwasglue/" target="_blank">gwasglue</a> R package aims to extend the use of these data to further types of statistical methods.</p>
                    </section>

                    <hr class='my-4'>

                    <section>
                        <h3 id="authentication">Authentication</h3>
                        <table class="table table-sm">
                            <thead><tr><td><b>Endpoint</b></td><td width="70%"><b>Authentication</b></td></tr></thead>
                            <tbody>
                                <tr><td>/status, /batches</td><td>None</td></tr>
                                <tr><td>All other endpoints</td><td>JWT (required)</td></tr>
                            </tbody>
                        </table>

                        <hr><h5>JSON Web Token (JWT)</h5>
                        <p>You need to prove your identity (authenticate) to use our service, even if you are querying a public dataset.</p>
                        <p>JSON Web Token (JWT) is used for authentication. The token is required for most endpoints. To obtain a token, sign in to the <a href="{{ root_url }}/profile">account</a> page. You will be prompted to generate a token which is valid for {{ jwt_validity }} days.</p>

                        <hr><h5>If you are using packages developed by us</h5>
                        <p>The ieugwasr and ieugwaspy packages are updated regularly to keep abreast of the changes of OpenGWAS API. If you use ieugwasr or TwoSampleMR etc. (which depends on ieugwasr), read <a href="https://mrcieu.github.io/ieugwasr/articles/guide.html#authentication" target="_blank">ieugwasr tutorial</a> about how to set up JWT. For ieugwaspy, refer to the  <a href="https://mrcieu.github.io/ieugwaspy/api.html" target="_blank">ieugwaspy docs</a>.</p>

                        <hr><h5>If you are building you own packages</h5>
                        <p>The <a href="{{ url_for('api.doc') }}" target="_blank">API documentation</a> provides a list of endpoints, whether they are protected (require authentication), the parameters and responses, and an interactive way to try out. To try the protected endpoints, click the green 'Authorize' button in API docs and provide your token as instructed.</p>
                        <p><u><i>IMPORTANT - In either circumstances below you must cite OpenGWAS properly - OpenGWAS must not be provided as "white label" software or service by any means.</i></u></p>
                        <p>(1) If you package depends on ieugwasr or ieugwaspy, make sure your end users are aware of OpenGWAS JWT, preferably by including a link to this page in your product docs.</p>
                        <p>(2) If you are building from scratch i.e. writing your own code to make API calls to OpenGWAS, in addition to (1), you also need to implement something that adds JWT into the request header under the key <code>Authorization</code> with value <code>Bearer <i>your_token</i></code>. Note that there is a whitespace between <code>Bearer</code> and your token.</p>
                        <p>You may want to visit the <a href="{{ url_for('api.doc') }}" target="_blank">API docs</a> and find the <code>/user</code> endpoint, which is created to help you validate your implementation of the authentication process.
                            <br>- <b>What will happen without a valid token?</b><br>If you just click 'Try it out' then 'Execute', you will get a '401 Unauthorized' error.
                            <br>- <b>What if I have a valid token?</b><br>If you focus on the upper right corner of the /user area, you will see a lock icon. Click the lock icon and you will find 'token_jwt' as available authorizations. Follow the instruction, click 'Authorize', and finally try to execute on the /user endpoint again. This time you will get '200 OK' and you can see your user information in the response body.
                            <br>- <b>Is there any other error?</b><br>Yes. You may see other errors (e.g. if you go beyond the rate limit you will get the '429 Too Many Requests' error), so when you are using the 200 status code as the indication of success, you may also want to check the response body.</p>
                    </section>

                    <hr class='my-4'>

                    <section>
                        <h3 id="allowance">Allowance and cost</h3>
                        <h5>Allowance (free credits)</h5>
                        <p>As part of our fair use policy, users will be given allowance for the cost of the requests they make in each time span. The allowance is provided <u>free of charge</u>.</p>
                        <p>The allowance is reset periodically and the clock starts at your first request. E.g. if the allowance is "1000 per 10 minutes" and you made your first request at 09:41:19, your allowance will be reset to 1000 at 09:51:19. Your real-time allowance information is available on your <a href="{{ root_url }}/profile">account</a> page, or in the response header (X-Allowance-* and Retry-After) of every request, except for those endpoints of which the cost is 0.</p>
                        <p><u>Resetting your token (JWT) will not reset your allowance.</u> If you have run out of allowance, you have to wait until the next period. If you insist sending requests and accumulate '429 Too Many Requests' errors, you account and IP address may be blocked for up to a week.</p>
                        <p>What's my allowance?</p>
                        <p>After your first sign-in you will be in the 'Trial' tier which only offers limited allowance. Please follow the instructions on you account page to upgrade you account to Standard.</p>
                        <table class="table table-sm">
                            <thead><tr><td><b>User tier</b></td><td><b>Allowance (free credits)</b></td></tr></thead>
                            <tbody>
                                {% for key, name in user_tiers.items() if key != 'UOB' %}
                                    <tr><td>{{ name }}</td><td>{{ user_allowance[key] }}</td></tr>
                                {% endfor %}
                            </tbody>
                        </table>
                        <p><u>Note</u>: University of Bristol users (who signed in using Microsoft, GitHub or email verification, as long as the email address on your OpenGWAS profile ends with @bristol.ac.uk) will be given higher allowance than displayed above. Please check your <a href="{{ root_url }}/profile">account</a> page for details.</p>
                        <p>Read our <a href="https://blog.opengwas.io/posts/user-auth-spring-2024/" target="_blank">blog post</a> for more info.</p>
                        <h5>Cost (per request)</h5>
                        <p><table class="table table-sm">
                            <thead><tr><td><b>Endpoint</b></td><td width="40%"><b>Cost per request</b></td><td><b>Conditions and notes</b></td></tr></thead>
                            <tbody>
                                <tr><td>/status, /batches, /users</td><td>0 (no cost, but still subject to fair-use policy)</td><td></td></tr>
                                <tr><td>/gwasinfo<br><br>/gwasinfo/files</td><td>1<br>50<br>N<sub>id</sub> &times; 1000</td><td>1 &le; N<sub>id</sub> &le; 100<br>100 &lt; N<sub>id</sub> or no id is specified</td></tr>
                                <tr><td>/associations</td><td>max(N<sub>id</sub>, N<sub>variant</sub>)<br>max(N<sub>id</sub>, N<sub>variant</sub>) &times; 5</td><td><br>proxies = 1</td></tr>
                                <tr><td>/tophits</td><td>N<sub>id</sub><br>N<sub>id</sub> &times; 10<br>N<sub>id</sub> &times; 30</td><td>clump = 1 (use preclumped data)<br>clump = 0 (return all records)<br>clump = 1 (requires new clumping)</td></tr>
                                <tr><td>/ld/clump<br>/ld/matrix<br>/ld/reflookup</td><td>12<br>18<br>2</td><td></td></tr>
                                <tr><td>/phewas</td><td>(N<sub>variant_rsid</sub> + N<sub>variant_chrpos</sub> + N<sub>variant_cprange</sub>) &times; 20 + N<sub>results</sub></td><td></td></tr>
                                <tr><td>/variants<br>/variants/afl2/snplist</td><td>1<br>10</td><td></td></tr>
                            </tbody>
                        </table></p>
                    </section>

                    <hr class='my-4'>

                    <section>
                        <h3 id="tools">Related tools</h3>
                        <p>For ease of use we have developed two packages which are solely designed to create easy interfaces to the API endpoints:</p>
                        <ul>
                            <li>The ieugwasr R package <a href="https://mrcieu.github.io/ieugwasr">[link]</a></li>
                            <li>The ieugwaspy python package <a href="https://github.com/mrcieu/ieugwaspy">[link]</a></li>
                        </ul>
                        <p>The following packages also connect to these data:</p>
                        <ul>
                            <li>The TwoSampleMR R package automatically performs Mendelian randomization on the data accessed through the API <a href="https://github.com/mrcieu/TwoSampleMR">[link]</a></li>
                            <li>The gwasglue R packages connects the data to several other analytical tools related to e.g. colocalisation, fine mapping, etc <a href="https://mrcieu.github.io/gwasglue/">[link]</a></li>
                            <li>The <a href="https://gwas.mrcieu.ac.uk">https://gwas.mrcieu.ac.uk</a> page provides an interactive way to browse through the available datasets.</li>
                            <li>The MR-Base web app uses the underlying data for basic interactive <a href="http://www.mrbase.org">[link]</a></li>
                        </ul>
                    </section>
                </div>
            </div>

      </div>

    </main>

    <hr class='my-4'>

    <script src="https://cdnjs.cloudflare.com/ajax/libs/tocbot/4.23.0/tocbot.min.js" integrity="sha512-83MzddG6eG9xvFY5tqqGsjQ+g1pHOK1/nu5aI5oR8rnw43wWtzXBS+LyJTbkhcNwRuL2z/HuNdeotr5Q/rmFwA==" crossorigin="anonymous" referrerpolicy="no-referrer"></script>
    <script>
        // https://github.com/tscanlin/tocbot
        tocbot.init({
            tocSelector: '.js-toc',
            contentSelector: '.js-toc-content',
            headingSelector: 'h3',
            hasInnerContainers: true,
            orderedList: false
        });
    </script>
</body>
</html>
